File: Fortran/2_pingPong/solution/a/pingPong.f90
program pingPong
    use mpi
    implicit none

    ! Variables for the process rank and number of processes
    integer :: myRank, numProcs, ierr
    integer :: pingCount, pongCount

    pingCount = 42
    pongCount = 0

    ! Initialize MPI
    call MPI_INIT(ierr)

    ! Find out MPI communicator size and process rank
    call MPI_COMM_RANK(MPI_COMM_WORLD, myRank, ierr)
    call MPI_COMM_SIZE(MPI_COMM_WORLD, numProcs, ierr)

    ! Have only the first process execute the following code
    if (myRank == 0) then
        print '("Sending Ping (# ",I0,")")', pingCount
        ! Send pingCount
        call MPI_SEND(pingCount, 1, MPI_INTEGER, 1, 0, &
            MPI_COMM_WORLD, ierr)

        ! Receive pongCount
        call MPI_RECV(pongCount, 1, MPI_INTEGER, 1, 0, &
            MPI_COMM_WORLD, MPI_STATUS_IGNORE, ierr)
        
        print '("Received Pong (# ",I0,")")', pongCount
    else
        ! Do proper receive and send in other process

        ! Receive pingCount
        call MPI_RECV(pingCount, 1, MPI_INTEGER, 0, 0, &
            MPI_COMM_WORLD, MPI_STATUS_IGNORE, ierr)

        print '("Received Ping (# ",I0,")")', pingCount

        ! Calculate and send pongCount
        pongCount = -pingCount
        call MPI_SEND(pongCount, 1, MPI_INTEGER, 0, 0, &
            MPI_COMM_WORLD, ierr)

        print '("Sending Pong (# ",I0,")")', pongCount
    end if

    ! Finialize MPI
    call MPI_FINALIZE(ierr)

end program pingPong

File: C/2_pingPong/solution/a/pingPong.c
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>

int main(int argc, char **argv)
{
    int worldRank, worldSize;
    int pingCount = 42;
    int pongCount = 0;

    // Initialize MPI
    MPI_Init(&argc, &argv);

    // Find out MPI communicator size and process rank
    MPI_Comm_size(MPI_COMM_WORLD, &worldSize);
    MPI_Comm_rank(MPI_COMM_WORLD, &worldRank);

    // Have only the first process execute the following code
    if (worldRank == 0)
    {
        printf("Sending Ping (# %i)\n", pingCount);

        // Send pingCount
        MPI_Send(&pingCount, 1, MPI_INT, 1, 0, MPI_COMM_WORLD);

        // Receive pongCount
        MPI_Recv(&pongCount, 1, MPI_INT, 1, 0, MPI_COMM_WORLD,
            MPI_STATUS_IGNORE);

        printf("Received Pong (# %i)\n", pongCount);
    }
    // Do proper receive and send in any other process
    else
    {
        // Receive pingCount
        MPI_Recv(&pingCount, 1, MPI_INT, 0, 0, MPI_COMM_WORLD,
            MPI_STATUS_IGNORE);

        printf("Received Ping (# %i)\n", pingCount);

        // calculate and send pongCount
        pongCount -= pingCount;
        printf("Sending Pong (# %i)\n", pongCount);
        MPI_Send(&pongCount, 1, MPI_INT, 0, 0, MPI_COMM_WORLD);
    }

    // Finalize MPI
    MPI_Finalize();

    return EXIT_SUCCESS;
}
